{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "outputs": [],
   "source": [
    "from Bio import pairwise2\n",
    "from Bio.Seq import Seq\n",
    "S =#%%\n",
    "from Bio import pairwise2\n",
    "from Bio.Seq import Seq\n",
    "S = '''\n",
    "from Bio import SubsMat\n",
    "\n",
    "# Define the weights for the alignment\n",
    "weights = MatrixInfo.blosum62\n",
    "\n",
    "# Compute the pairwise alignments and scores\n",
    "for i in range(len(sequences)):\n",
    "    for j in range(i+1, len(sequences)):\n",
    "        seq1 = sequences[i]\n",
    "        seq2 = sequences[j]\n",
    "        alignment = pairwise2.align.globalds(seq1, seq2, weights, -10, -0.5)\n",
    "        score = alignment[0][2]\n",
    "        print(f\"Alignment score between {seq1} and {seq2}: {score}\")\n"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "outputs": [],
   "source": [
    "'''\n",
    "\n",
    "# Example TRAV gene segments\n",
    "seq1 = \"26-1\"\n",
    "seq2 = \"26-2\"\n",
    "\n",
    "# Convert the sequences to Biopython Seq objects\n",
    "seq1 = Seq(seq1)\n",
    "seq2 = Seq(seq2)\n",
    "\n",
    "# Align the sequences using the Needleman-Wunsch algorithm\n",
    "alignments = pairwise2.align.globalxx(seq1, seq2)\n",
    "\n",
    "# Print the alignment with the highest score\n",
    "print(pairwise2.format_alignment(*alignments[0]))"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "outputs": [],
   "source": [
    "from Bio import pairwise2\n",
    "\n",
    "# Example TRAV gene segments\n",
    "sequences = [\"TRAV26-1*01\", \"TRAV20*01\", \"TRAV38-2/DV8*01\", \"TRAV26-1*01\", \"TRAV20*01\", \"TRAV26-1*01\", \"TRAV20*01\", \"TRAV26-1*01\", \"TRAV20*01\", \"TRAV38-1*01\", \"TRAV12-2*01\", \"TRAV12-2*01\", \"TRAV12-2*01\", \"TRAV12-2*01\", \"TRAV12-2*01\", \"TRAV12-2*01\", \"TRAV12-2*01\", \"TRAV12-2*01\", \"TRAV12-2*01\", \"TRAV12-2*01\", \"TRAV12-2*01\", \"TRAV12-2*01\", \"TRAV12-2*01\", \"TRAV12-2*01\", \"TRAV12-2*01\", \"TRAV3*01\", \"TRAV12-1*01\", \"TRAV27*01\", \"TRAV1-2*01\", \"TRAV35*01\", \"TRAV13-2*01\", \"TRAV12-2*01\", \"TRAV12-2*01\", \"TRAV12-2*01\", \"TRAV12-2*01\", \"TRAV12-2*01\", \"TRAV12-2*01\", \"TRAV12-2*01\", \"TRAV12-2*01\", \"TRAV3*01\", \"TRAV12-2*01\", \"TRAV12-2*01\", \"TRAV12-2*01\", \"TRAV12-2*01\", \"TRAV12-2*01\", \"TRAV12-2*01\", \"TRAV12-2*01\", \"TRAV12-3*01\", \"TRAV12-3*01\", \"TRAV12-3*01\", \"TRAV13-1*01\", \"TRAV13-1*01\", \"TRAV13-1*01\", \"TRAV13-1*01\", \"TRAV13-1*01\", \"TRAV13-1*01\", \"TRAV8-4*01\", \"TRAV8-1*01\", \"TRAV3*01\", \"TRAV14/DV4*01\", \"TRAV3*01\", \"TRAV26-1*01\"]\n",
    "\n",
    "\n",
    "# Compute pairwise alignment scores using Needleman-Wunsch\n",
    "align_scores = {}\n",
    "for i, seq1 in enumerate(sequences):\n",
    "    for j, seq2 in enumerate(sequences):\n",
    "        if j > i:\n",
    "            align_score = pairwise2.align.globalxx(seq1, seq2, score_only=True)\n",
    "            align_scores[(i, j)] = align_score\n",
    "            align_scores[(j, i)] = align_score\n",
    "\n",
    "# Print the alignment scores\n",
    "for i, seq1 in enumerate(sequences):\n",
    "    for j, seq2 in enumerate(sequences):\n",
    "        if j > i:\n",
    "            print(f\"Alignment score between {seq1} and {seq2}: {align_scores[(i, j)]}\")"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "outputs": [],
   "source": [
    "from Bio import pairwise2\n",
    "\n",
    "def get_similarity_score(seq1, seq2):\n",
    "    alignments = pairwise2.align.globalxx(seq1, seq2)\n",
    "    alignment_scores = [a[2] for a in alignments]\n",
    "    return max(alignment_scores, default=0)\n",
    "\n",
    "sequences = [\"26-1*01\", \"TRAV20*01\", \"TRAV38-2/DV8*01\", \"TRAV26-1*01\", \"TRAV20*01\", \"TRAV26-1*01\", \"TRAV20*01\", \"TRAV26-1*01\", \"TRAV20*01\", \"TRAV38-1*01\", \"TRAV12-2*01\", \"TRAV12-2*01\", \"TRAV12-2*01\", \"TRAV12-2*01\", \"TRAV12-2*01\", \"TRAV12-2*01\", \"TRAV12-2*01\", \"TRAV12-2*01\", \"TRAV12-2*01\", \"TRAV12-2*01\", \"TRAV12-2*01\", \"TRAV12-2*01\", \"TRAV12-2*01\", \"TRAV12-2*01\", \"TRAV12-2*01\", \"TRAV3*01\", \"TRAV12-1*01\", \"TRAV27*01\", \"TRAV1-2*01\", \"TRAV35*01\", \"TRAV13-2*01\", \"TRAV12-2*01\", \"TRAV12-2*01\", \"TRAV12-2*01\", \"TRAV12-2*01\", \"TRAV12-2*01\", \"TRAV12-2*01\", \"TRAV12-2*01\", \"TRAV12-2*01\", \"TRAV3*01\", \"TRAV12-2*01\", \"TRAV12-2*01\", \"TRAV12-2*01\", \"TRAV12-2*01\", \"TRAV12-2*01\", \"TRAV12-2*01\", \"TRAV12-2*01\", \"TRAV12-3*01\", \"TRAV12-3*01\", \"TRAV12-3*01\", \"TRAV13-1*01\", \"TRAV13-1*01\", \"TRAV13-1*01\", \"TRAV13-1*01\", \"TRAV13-1*01\", \"TRAV13-1*01\", \"TRAV8-4*01\", \"TRAV8-1*01\", \"TRAV3*01\", \"TRAV14/DV4*01\", \"TRAV3*01\", \"TRAV26-1*01\"]\n",
    "\n",
    "scores = {}\n",
    "for seq in sequences:\n",
    "    score = sum(get_similarity_score(seq, other_seq) for other_seq in sequences)\n",
    "    scores[seq] = score\n",
    "\n",
    "print(scores)\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "outputs": [],
   "source": [
    "'''\n",
    "from sklearn.feature_extraction.text import CountVectorizer\n",
    "from sklearn.metrics.pairwise import pairwise_distances\n",
    "import scipy.spatial.distance as dist\n",
    "\n",
    "# Create a count vectorizer to convert the strings into binary term frequencies\n",
    "vectorizer = CountVectorizer(binary=True)\n",
    "X = vectorizer.fit_transform(sequences)\n",
    "\n",
    "dense_matrix = X.toarray()\n",
    "\n",
    "# Compute pairwise distances between rows\n",
    "distances = dist.pdist(dense_matrix, metric='cosine')\n",
    "square_distances = dist.squareform(distances)\n",
    "\n",
    "# Calculate the score for each string as the average similarity with the other strings\n",
    "scores = []\n",
    "for i in range(len(sequences)):\n",
    "    score = sum(square_distances[i]) / (len(sequences) - 1)\n",
    "    scores.append(score)\n",
    "\n",
    "# Print the scores for each string\n",
    "for i in range(len(sequences)):\n",
    "    print(f\"{sequences[i]}: {scores[i]}\")\n",
    "'''"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "outputs": [],
   "source": [
    "from src.preprocess.setup import config\n",
    "from src.preprocess.feature_filter import load_data\n",
    "from src.preprocess.label import generate_label\n",
    "import src.preprocess.feature_filter\n",
    "\n",
    "config.setConfig(config.speciesType.human, config.chainType.alpha)\n",
    "data = load_data()"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "outputs": [],
   "source": [
    "sequences_1 = data['v_a_gene']\n",
    "print(sequences_1)"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "outputs": [],
   "source": [
    "\n",
    "from Bio.Align.Applications import MuscleCommandline\n",
    "from io import StringIO\n",
    "from Bio import AlignIO\n",
    "\n",
    "# Extract V column from sequences list\n",
    "sequences_V = [s.split('\\t')[0] for s in sequences]\n",
    "\n",
    "# Write sequences to a temporary file\n",
    "temp_seq_file = 'sequences.fasta'\n",
    "with open(temp_seq_file, 'w') as f:\n",
    "    for i, seq in enumerate(sequences_V):\n",
    "        f.write('>seq{}\\n{}\\n'.format(i, seq))\n",
    "\n",
    "# Run MUSCLE alignment\n",
    "muscle_exe = '/Users/liuqianyi/Documents/muscle/muscle.macos_intel64'\n",
    "muscle_cline = MuscleCommandline(muscle_exe, input=temp_seq_file)\n",
    "stdout, stderr = muscle_cline()\n",
    "align = AlignIO.read(StringIO(stdout), 'fasta')\n",
    "\n",
    "# Print similarity scores\n",
    "for i, record1 in enumerate(align):\n",
    "    for j, record2 in enumerate(align):\n",
    "        score = align[:, i].similarity(align[:, j])\n",
    "        print('{} vs {}: {}'.format(record1.id, record2.id, score))\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "outputs": [],
   "source": [],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "5 ATCG\n",
      "  ||||\n",
      "1 ATCG\n",
      "  Score=4\n",
      "\n"
     ]
    }
   ],
   "source": [
    "\n",
    "# Example TRAV gene segments\n",
    "seq1 = \"BYJEATCG\"\n",
    "seq2 = \"ATCG\"\n",
    "\n",
    "# Convert the sequences to Biopython Seq objects\n",
    "seq1 = Seq(seq1)\n",
    "seq2 = Seq(seq2)\n",
    "\n",
    "\n",
    "# Align the sequences using the Needleman-Wunsch algorithm\n",
    "alignments = pairwise2.align.localms(seq1, seq2, 1, -1, -0.5, -0.1, one_alignment_only=True)\n",
    "\n",
    "# Print the alignment with the highest score\n",
    "print(pairwise2.format_alignment(*alignments[0]))"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Alignment score between 12-2 and 12-1: 3.0\n",
      "Alignment score between 12-2 and 12-3: 3.0\n",
      "Alignment score between 12-2 and 13-1: 2.0\n",
      "Alignment score between 12-2 and 8-4: 1.0\n",
      "Alignment score between 12-2 and 8-1: 1.0\n",
      "Alignment score between 12-2 and 14/DV4: 1.0\n",
      "Alignment score between 12-2 and 4: 0.0\n",
      "Alignment score between 12-2 and 5: 0.0\n",
      "Alignment score between 12-2 and 19: 1.0\n",
      "Alignment score between 12-2 and 9-2: 2.0\n",
      "Alignment score between 12-2 and 22: 2.0\n",
      "Alignment score between 12-2 and 25: 1.0\n",
      "Alignment score between 12-2 and 10: 1.0\n",
      "Alignment score between 12-2 and 39: 0.0\n",
      "Alignment score between 12-2 and 41: 1.0\n",
      "Alignment score between 12-2 and 2: 1.0\n",
      "Alignment score between 12-2 and 9-1: 1.0\n",
      "Alignment score between 12-2 and 1-1: 2.0\n",
      "Alignment score between 12-2 and 8-6: 1.0\n",
      "Alignment score between 12-2 and 26-2: 3.0\n",
      "Alignment score between 12-2 and 24: 1.0\n",
      "Alignment score between 12-2 and 8-3: 1.0\n",
      "Alignment score between 12-2 and 17: 1.0\n",
      "Alignment score between 12-2 and 29/DV5: 1.0\n",
      "Alignment score between 12-2 and 16: 1.0\n",
      "Alignment score between 12-2 and 30: 0.0\n",
      "Alignment score between 12-2 and 6: 0.0\n",
      "Alignment score between 12-2 and 23/DV6: 1.0\n",
      "Alignment score between 12-2 and 36/DV7: 0.0\n",
      "Alignment score between 12-2 and 8-2: 2.0\n",
      "Alignment score between 12-2 and 18: 1.0\n",
      "Alignment score between 12-2 and 34: 0.0\n",
      "Alignment score between 12-2 and 40: 0.0\n",
      "Alignment score between 12-2 and 7: 0.0\n",
      "Alignment score between 12-2 and 14/DV4*02: 2.0\n",
      "Alignment score between 12-2 and 8-6*02: 2.0\n",
      "Alignment score between 12-1 and 12-3: 3.0\n",
      "Alignment score between 12-1 and 13-1: 3.0\n",
      "Alignment score between 12-1 and 8-4: 1.0\n",
      "Alignment score between 12-1 and 8-1: 2.0\n",
      "Alignment score between 12-1 and 14/DV4: 1.0\n",
      "Alignment score between 12-1 and 4: 0.0\n",
      "Alignment score between 12-1 and 5: 0.0\n",
      "Alignment score between 12-1 and 19: 1.0\n",
      "Alignment score between 12-1 and 9-2: 1.0\n",
      "Alignment score between 12-1 and 22: 1.0\n",
      "Alignment score between 12-1 and 25: 1.0\n",
      "Alignment score between 12-1 and 10: 1.0\n",
      "Alignment score between 12-1 and 39: 0.0\n",
      "Alignment score between 12-1 and 41: 1.0\n",
      "Alignment score between 12-1 and 2: 1.0\n",
      "Alignment score between 12-1 and 9-1: 2.0\n",
      "Alignment score between 12-1 and 1-1: 3.0\n",
      "Alignment score between 12-1 and 8-6: 1.0\n",
      "Alignment score between 12-1 and 26-2: 2.0\n",
      "Alignment score between 12-1 and 24: 1.0\n",
      "Alignment score between 12-1 and 8-3: 1.0\n",
      "Alignment score between 12-1 and 17: 1.0\n",
      "Alignment score between 12-1 and 29/DV5: 1.0\n",
      "Alignment score between 12-1 and 16: 1.0\n",
      "Alignment score between 12-1 and 30: 0.0\n",
      "Alignment score between 12-1 and 6: 0.0\n",
      "Alignment score between 12-1 and 23/DV6: 1.0\n",
      "Alignment score between 12-1 and 36/DV7: 0.0\n",
      "Alignment score between 12-1 and 8-2: 1.0\n",
      "Alignment score between 12-1 and 18: 1.0\n",
      "Alignment score between 12-1 and 34: 0.0\n",
      "Alignment score between 12-1 and 40: 0.0\n",
      "Alignment score between 12-1 and 7: 0.0\n",
      "Alignment score between 12-1 and 14/DV4*02: 2.0\n",
      "Alignment score between 12-1 and 8-6*02: 1.0\n",
      "Alignment score between 12-3 and 13-1: 2.0\n",
      "Alignment score between 12-3 and 8-4: 1.0\n",
      "Alignment score between 12-3 and 8-1: 1.0\n",
      "Alignment score between 12-3 and 14/DV4: 1.0\n",
      "Alignment score between 12-3 and 4: 0.0\n",
      "Alignment score between 12-3 and 5: 0.0\n",
      "Alignment score between 12-3 and 19: 1.0\n",
      "Alignment score between 12-3 and 9-2: 1.0\n",
      "Alignment score between 12-3 and 22: 1.0\n",
      "Alignment score between 12-3 and 25: 1.0\n",
      "Alignment score between 12-3 and 10: 1.0\n",
      "Alignment score between 12-3 and 39: 1.0\n",
      "Alignment score between 12-3 and 41: 1.0\n",
      "Alignment score between 12-3 and 2: 1.0\n",
      "Alignment score between 12-3 and 9-1: 1.0\n",
      "Alignment score between 12-3 and 1-1: 2.0\n",
      "Alignment score between 12-3 and 8-6: 1.0\n",
      "Alignment score between 12-3 and 26-2: 2.0\n",
      "Alignment score between 12-3 and 24: 1.0\n",
      "Alignment score between 12-3 and 8-3: 2.0\n",
      "Alignment score between 12-3 and 17: 1.0\n",
      "Alignment score between 12-3 and 29/DV5: 1.0\n",
      "Alignment score between 12-3 and 16: 1.0\n",
      "Alignment score between 12-3 and 30: 1.0\n",
      "Alignment score between 12-3 and 6: 0.0\n",
      "Alignment score between 12-3 and 23/DV6: 2.0\n",
      "Alignment score between 12-3 and 36/DV7: 1.0\n",
      "Alignment score between 12-3 and 8-2: 1.0\n",
      "Alignment score between 12-3 and 18: 1.0\n",
      "Alignment score between 12-3 and 34: 1.0\n",
      "Alignment score between 12-3 and 40: 0.0\n",
      "Alignment score between 12-3 and 7: 0.0\n",
      "Alignment score between 12-3 and 14/DV4*02: 2.0\n",
      "Alignment score between 12-3 and 8-6*02: 1.0\n",
      "Alignment score between 13-1 and 8-4: 1.0\n",
      "Alignment score between 13-1 and 8-1: 2.0\n",
      "Alignment score between 13-1 and 14/DV4: 1.0\n",
      "Alignment score between 13-1 and 4: 0.0\n",
      "Alignment score between 13-1 and 5: 0.0\n",
      "Alignment score between 13-1 and 19: 1.0\n",
      "Alignment score between 13-1 and 9-2: 1.0\n",
      "Alignment score between 13-1 and 22: 0.0\n",
      "Alignment score between 13-1 and 25: 0.0\n",
      "Alignment score between 13-1 and 10: 1.0\n",
      "Alignment score between 13-1 and 39: 1.0\n",
      "Alignment score between 13-1 and 41: 1.0\n",
      "Alignment score between 13-1 and 2: 0.0\n",
      "Alignment score between 13-1 and 9-1: 2.0\n",
      "Alignment score between 13-1 and 1-1: 3.0\n",
      "Alignment score between 13-1 and 8-6: 1.0\n",
      "Alignment score between 13-1 and 26-2: 1.0\n",
      "Alignment score between 13-1 and 24: 0.0\n",
      "Alignment score between 13-1 and 8-3: 1.0\n",
      "Alignment score between 13-1 and 17: 1.0\n",
      "Alignment score between 13-1 and 29/DV5: 0.0\n",
      "Alignment score between 13-1 and 16: 1.0\n",
      "Alignment score between 13-1 and 30: 1.0\n",
      "Alignment score between 13-1 and 6: 0.0\n",
      "Alignment score between 13-1 and 23/DV6: 1.0\n",
      "Alignment score between 13-1 and 36/DV7: 1.0\n",
      "Alignment score between 13-1 and 8-2: 1.0\n",
      "Alignment score between 13-1 and 18: 1.0\n",
      "Alignment score between 13-1 and 34: 1.0\n",
      "Alignment score between 13-1 and 40: 0.0\n",
      "Alignment score between 13-1 and 7: 0.0\n",
      "Alignment score between 13-1 and 14/DV4*02: 1.0\n",
      "Alignment score between 13-1 and 8-6*02: 1.0\n",
      "Alignment score between 8-4 and 8-1: 2.0\n",
      "Alignment score between 8-4 and 14/DV4: 1.0\n",
      "Alignment score between 8-4 and 4: 1.0\n",
      "Alignment score between 8-4 and 5: 0.0\n",
      "Alignment score between 8-4 and 19: 0.0\n",
      "Alignment score between 8-4 and 9-2: 1.0\n",
      "Alignment score between 8-4 and 22: 0.0\n",
      "Alignment score between 8-4 and 25: 0.0\n",
      "Alignment score between 8-4 and 10: 0.0\n",
      "Alignment score between 8-4 and 39: 0.0\n",
      "Alignment score between 8-4 and 41: 1.0\n",
      "Alignment score between 8-4 and 2: 0.0\n",
      "Alignment score between 8-4 and 9-1: 1.0\n",
      "Alignment score between 8-4 and 1-1: 1.0\n",
      "Alignment score between 8-4 and 8-6: 2.0\n",
      "Alignment score between 8-4 and 26-2: 1.0\n",
      "Alignment score between 8-4 and 24: 1.0\n",
      "Alignment score between 8-4 and 8-3: 2.0\n",
      "Alignment score between 8-4 and 17: 0.0\n",
      "Alignment score between 8-4 and 29/DV5: 0.0\n",
      "Alignment score between 8-4 and 16: 0.0\n",
      "Alignment score between 8-4 and 30: 0.0\n",
      "Alignment score between 8-4 and 6: 0.0\n",
      "Alignment score between 8-4 and 23/DV6: 0.0\n",
      "Alignment score between 8-4 and 36/DV7: 0.0\n",
      "Alignment score between 8-4 and 8-2: 2.0\n",
      "Alignment score between 8-4 and 18: 1.0\n",
      "Alignment score between 8-4 and 34: 1.0\n",
      "Alignment score between 8-4 and 40: 1.0\n",
      "Alignment score between 8-4 and 7: 0.0\n",
      "Alignment score between 8-4 and 14/DV4*02: 1.0\n",
      "Alignment score between 8-4 and 8-6*02: 2.0\n",
      "Alignment score between 8-1 and 14/DV4: 1.0\n",
      "Alignment score between 8-1 and 4: 0.0\n",
      "Alignment score between 8-1 and 5: 0.0\n",
      "Alignment score between 8-1 and 19: 1.0\n",
      "Alignment score between 8-1 and 9-2: 1.0\n",
      "Alignment score between 8-1 and 22: 0.0\n",
      "Alignment score between 8-1 and 25: 0.0\n",
      "Alignment score between 8-1 and 10: 1.0\n",
      "Alignment score between 8-1 and 39: 0.0\n",
      "Alignment score between 8-1 and 41: 1.0\n",
      "Alignment score between 8-1 and 2: 0.0\n",
      "Alignment score between 8-1 and 9-1: 2.0\n",
      "Alignment score between 8-1 and 1-1: 2.0\n",
      "Alignment score between 8-1 and 8-6: 2.0\n",
      "Alignment score between 8-1 and 26-2: 1.0\n",
      "Alignment score between 8-1 and 24: 0.0\n",
      "Alignment score between 8-1 and 8-3: 2.0\n",
      "Alignment score between 8-1 and 17: 1.0\n",
      "Alignment score between 8-1 and 29/DV5: 0.0\n",
      "Alignment score between 8-1 and 16: 1.0\n",
      "Alignment score between 8-1 and 30: 0.0\n",
      "Alignment score between 8-1 and 6: 0.0\n",
      "Alignment score between 8-1 and 23/DV6: 0.0\n",
      "Alignment score between 8-1 and 36/DV7: 0.0\n",
      "Alignment score between 8-1 and 8-2: 2.0\n",
      "Alignment score between 8-1 and 18: 1.0\n",
      "Alignment score between 8-1 and 34: 0.0\n",
      "Alignment score between 8-1 and 40: 0.0\n",
      "Alignment score between 8-1 and 7: 0.0\n",
      "Alignment score between 8-1 and 14/DV4*02: 1.0\n",
      "Alignment score between 8-1 and 8-6*02: 2.0\n",
      "Alignment score between 14/DV4 and 4: 1.0\n",
      "Alignment score between 14/DV4 and 5: 0.0\n",
      "Alignment score between 14/DV4 and 19: 1.0\n",
      "Alignment score between 14/DV4 and 9-2: 0.0\n",
      "Alignment score between 14/DV4 and 22: 0.0\n",
      "Alignment score between 14/DV4 and 25: 0.0\n",
      "Alignment score between 14/DV4 and 10: 1.0\n",
      "Alignment score between 14/DV4 and 39: 0.0\n",
      "Alignment score between 14/DV4 and 41: 1.0\n",
      "Alignment score between 14/DV4 and 2: 0.0\n",
      "Alignment score between 14/DV4 and 9-1: 1.0\n",
      "Alignment score between 14/DV4 and 1-1: 1.0\n",
      "Alignment score between 14/DV4 and 8-6: 0.0\n",
      "Alignment score between 14/DV4 and 26-2: 0.0\n",
      "Alignment score between 14/DV4 and 24: 1.0\n",
      "Alignment score between 14/DV4 and 8-3: 0.0\n",
      "Alignment score between 14/DV4 and 17: 1.0\n",
      "Alignment score between 14/DV4 and 29/DV5: 3.0\n",
      "Alignment score between 14/DV4 and 16: 1.0\n",
      "Alignment score between 14/DV4 and 30: 0.0\n",
      "Alignment score between 14/DV4 and 6: 0.0\n",
      "Alignment score between 14/DV4 and 23/DV6: 3.0\n",
      "Alignment score between 14/DV4 and 36/DV7: 3.0\n",
      "Alignment score between 14/DV4 and 8-2: 0.0\n",
      "Alignment score between 14/DV4 and 18: 1.0\n",
      "Alignment score between 14/DV4 and 34: 1.0\n",
      "Alignment score between 14/DV4 and 40: 1.0\n",
      "Alignment score between 14/DV4 and 7: 0.0\n",
      "Alignment score between 14/DV4 and 14/DV4*02: 6.0\n",
      "Alignment score between 14/DV4 and 8-6*02: 0.0\n",
      "Alignment score between 4 and 5: 0.0\n",
      "Alignment score between 4 and 19: 0.0\n",
      "Alignment score between 4 and 9-2: 0.0\n",
      "Alignment score between 4 and 22: 0.0\n",
      "Alignment score between 4 and 25: 0.0\n",
      "Alignment score between 4 and 10: 0.0\n",
      "Alignment score between 4 and 39: 0.0\n",
      "Alignment score between 4 and 41: 1.0\n",
      "Alignment score between 4 and 2: 0.0\n",
      "Alignment score between 4 and 9-1: 0.0\n",
      "Alignment score between 4 and 1-1: 0.0\n",
      "Alignment score between 4 and 8-6: 0.0\n",
      "Alignment score between 4 and 26-2: 0.0\n",
      "Alignment score between 4 and 24: 1.0\n",
      "Alignment score between 4 and 8-3: 0.0\n",
      "Alignment score between 4 and 17: 0.0\n",
      "Alignment score between 4 and 29/DV5: 0.0\n",
      "Alignment score between 4 and 16: 0.0\n",
      "Alignment score between 4 and 30: 0.0\n",
      "Alignment score between 4 and 6: 0.0\n",
      "Alignment score between 4 and 23/DV6: 0.0\n",
      "Alignment score between 4 and 36/DV7: 0.0\n",
      "Alignment score between 4 and 8-2: 0.0\n",
      "Alignment score between 4 and 18: 0.0\n",
      "Alignment score between 4 and 34: 1.0\n",
      "Alignment score between 4 and 40: 1.0\n",
      "Alignment score between 4 and 7: 0.0\n",
      "Alignment score between 4 and 14/DV4*02: 1.0\n",
      "Alignment score between 4 and 8-6*02: 0.0\n",
      "Alignment score between 5 and 19: 0.0\n",
      "Alignment score between 5 and 9-2: 0.0\n",
      "Alignment score between 5 and 22: 0.0\n",
      "Alignment score between 5 and 25: 1.0\n",
      "Alignment score between 5 and 10: 0.0\n",
      "Alignment score between 5 and 39: 0.0\n",
      "Alignment score between 5 and 41: 0.0\n",
      "Alignment score between 5 and 2: 0.0\n",
      "Alignment score between 5 and 9-1: 0.0\n",
      "Alignment score between 5 and 1-1: 0.0\n",
      "Alignment score between 5 and 8-6: 0.0\n",
      "Alignment score between 5 and 26-2: 0.0\n",
      "Alignment score between 5 and 24: 0.0\n",
      "Alignment score between 5 and 8-3: 0.0\n",
      "Alignment score between 5 and 17: 0.0\n",
      "Alignment score between 5 and 29/DV5: 1.0\n",
      "Alignment score between 5 and 16: 0.0\n",
      "Alignment score between 5 and 30: 0.0\n",
      "Alignment score between 5 and 6: 0.0\n",
      "Alignment score between 5 and 23/DV6: 0.0\n",
      "Alignment score between 5 and 36/DV7: 0.0\n",
      "Alignment score between 5 and 8-2: 0.0\n",
      "Alignment score between 5 and 18: 0.0\n",
      "Alignment score between 5 and 34: 0.0\n",
      "Alignment score between 5 and 40: 0.0\n",
      "Alignment score between 5 and 7: 0.0\n",
      "Alignment score between 5 and 14/DV4*02: 0.0\n",
      "Alignment score between 5 and 8-6*02: 0.0\n",
      "Alignment score between 19 and 9-2: 1.0\n",
      "Alignment score between 19 and 22: 0.0\n",
      "Alignment score between 19 and 25: 0.0\n",
      "Alignment score between 19 and 10: 1.0\n",
      "Alignment score between 19 and 39: 1.0\n",
      "Alignment score between 19 and 41: 1.0\n",
      "Alignment score between 19 and 2: 0.0\n",
      "Alignment score between 19 and 9-1: 1.0\n",
      "Alignment score between 19 and 1-1: 1.0\n",
      "Alignment score between 19 and 8-6: 0.0\n",
      "Alignment score between 19 and 26-2: 0.0\n",
      "Alignment score between 19 and 24: 0.0\n",
      "Alignment score between 19 and 8-3: 0.0\n",
      "Alignment score between 19 and 17: 1.0\n",
      "Alignment score between 19 and 29/DV5: 1.0\n",
      "Alignment score between 19 and 16: 1.0\n",
      "Alignment score between 19 and 30: 0.0\n",
      "Alignment score between 19 and 6: 0.0\n",
      "Alignment score between 19 and 23/DV6: 0.0\n",
      "Alignment score between 19 and 36/DV7: 0.0\n",
      "Alignment score between 19 and 8-2: 0.0\n",
      "Alignment score between 19 and 18: 1.0\n",
      "Alignment score between 19 and 34: 0.0\n",
      "Alignment score between 19 and 40: 0.0\n",
      "Alignment score between 19 and 7: 0.0\n",
      "Alignment score between 19 and 14/DV4*02: 1.0\n",
      "Alignment score between 19 and 8-6*02: 0.0\n",
      "Alignment score between 9-2 and 22: 1.0\n",
      "Alignment score between 9-2 and 25: 1.0\n",
      "Alignment score between 9-2 and 10: 0.0\n",
      "Alignment score between 9-2 and 39: 1.0\n",
      "Alignment score between 9-2 and 41: 0.0\n",
      "Alignment score between 9-2 and 2: 1.0\n",
      "Alignment score between 9-2 and 9-1: 2.0\n",
      "Alignment score between 9-2 and 1-1: 1.0\n",
      "Alignment score between 9-2 and 8-6: 1.0\n",
      "Alignment score between 9-2 and 26-2: 2.0\n",
      "Alignment score between 9-2 and 24: 1.0\n",
      "Alignment score between 9-2 and 8-3: 1.0\n",
      "Alignment score between 9-2 and 17: 0.0\n",
      "Alignment score between 9-2 and 29/DV5: 1.0\n",
      "Alignment score between 9-2 and 16: 0.0\n",
      "Alignment score between 9-2 and 30: 0.0\n",
      "Alignment score between 9-2 and 6: 0.0\n",
      "Alignment score between 9-2 and 23/DV6: 1.0\n",
      "Alignment score between 9-2 and 36/DV7: 0.0\n",
      "Alignment score between 9-2 and 8-2: 2.0\n",
      "Alignment score between 9-2 and 18: 0.0\n",
      "Alignment score between 9-2 and 34: 0.0\n",
      "Alignment score between 9-2 and 40: 0.0\n",
      "Alignment score between 9-2 and 7: 0.0\n",
      "Alignment score between 9-2 and 14/DV4*02: 1.0\n",
      "Alignment score between 9-2 and 8-6*02: 2.0\n",
      "Alignment score between 22 and 25: 1.0\n",
      "Alignment score between 22 and 10: 0.0\n",
      "Alignment score between 22 and 39: 0.0\n",
      "Alignment score between 22 and 41: 0.0\n",
      "Alignment score between 22 and 2: 1.0\n",
      "Alignment score between 22 and 9-1: 0.0\n",
      "Alignment score between 22 and 1-1: 0.0\n",
      "Alignment score between 22 and 8-6: 0.0\n",
      "Alignment score between 22 and 26-2: 2.0\n",
      "Alignment score between 22 and 24: 1.0\n",
      "Alignment score between 22 and 8-3: 0.0\n",
      "Alignment score between 22 and 17: 0.0\n",
      "Alignment score between 22 and 29/DV5: 1.0\n",
      "Alignment score between 22 and 16: 0.0\n",
      "Alignment score between 22 and 30: 0.0\n",
      "Alignment score between 22 and 6: 0.0\n",
      "Alignment score between 22 and 23/DV6: 1.0\n",
      "Alignment score between 22 and 36/DV7: 0.0\n",
      "Alignment score between 22 and 8-2: 1.0\n",
      "Alignment score between 22 and 18: 0.0\n",
      "Alignment score between 22 and 34: 0.0\n",
      "Alignment score between 22 and 40: 0.0\n",
      "Alignment score between 22 and 7: 0.0\n",
      "Alignment score between 22 and 14/DV4*02: 1.0\n",
      "Alignment score between 22 and 8-6*02: 1.0\n",
      "Alignment score between 25 and 10: 0.0\n",
      "Alignment score between 25 and 39: 0.0\n",
      "Alignment score between 25 and 41: 0.0\n",
      "Alignment score between 25 and 2: 1.0\n",
      "Alignment score between 25 and 9-1: 0.0\n",
      "Alignment score between 25 and 1-1: 0.0\n",
      "Alignment score between 25 and 8-6: 0.0\n",
      "Alignment score between 25 and 26-2: 1.0\n",
      "Alignment score between 25 and 24: 1.0\n",
      "Alignment score between 25 and 8-3: 0.0\n",
      "Alignment score between 25 and 17: 0.0\n",
      "Alignment score between 25 and 29/DV5: 2.0\n",
      "Alignment score between 25 and 16: 0.0\n",
      "Alignment score between 25 and 30: 0.0\n",
      "Alignment score between 25 and 6: 0.0\n",
      "Alignment score between 25 and 23/DV6: 1.0\n",
      "Alignment score between 25 and 36/DV7: 0.0\n",
      "Alignment score between 25 and 8-2: 1.0\n",
      "Alignment score between 25 and 18: 0.0\n",
      "Alignment score between 25 and 34: 0.0\n",
      "Alignment score between 25 and 40: 0.0\n",
      "Alignment score between 25 and 7: 0.0\n",
      "Alignment score between 25 and 14/DV4*02: 1.0\n",
      "Alignment score between 25 and 8-6*02: 1.0\n",
      "Alignment score between 10 and 39: 0.0\n",
      "Alignment score between 10 and 41: 1.0\n",
      "Alignment score between 10 and 2: 0.0\n",
      "Alignment score between 10 and 9-1: 1.0\n",
      "Alignment score between 10 and 1-1: 1.0\n",
      "Alignment score between 10 and 8-6: 0.0\n",
      "Alignment score between 10 and 26-2: 0.0\n",
      "Alignment score between 10 and 24: 0.0\n",
      "Alignment score between 10 and 8-3: 0.0\n",
      "Alignment score between 10 and 17: 1.0\n",
      "Alignment score between 10 and 29/DV5: 0.0\n",
      "Alignment score between 10 and 16: 1.0\n",
      "Alignment score between 10 and 30: 1.0\n",
      "Alignment score between 10 and 6: 0.0\n",
      "Alignment score between 10 and 23/DV6: 0.0\n",
      "Alignment score between 10 and 36/DV7: 0.0\n",
      "Alignment score between 10 and 8-2: 0.0\n",
      "Alignment score between 10 and 18: 1.0\n",
      "Alignment score between 10 and 34: 0.0\n",
      "Alignment score between 10 and 40: 1.0\n",
      "Alignment score between 10 and 7: 0.0\n",
      "Alignment score between 10 and 14/DV4*02: 2.0\n",
      "Alignment score between 10 and 8-6*02: 1.0\n",
      "Alignment score between 39 and 41: 0.0\n",
      "Alignment score between 39 and 2: 0.0\n",
      "Alignment score between 39 and 9-1: 1.0\n",
      "Alignment score between 39 and 1-1: 0.0\n",
      "Alignment score between 39 and 8-6: 0.0\n",
      "Alignment score between 39 and 26-2: 0.0\n",
      "Alignment score between 39 and 24: 0.0\n",
      "Alignment score between 39 and 8-3: 1.0\n",
      "Alignment score between 39 and 17: 0.0\n",
      "Alignment score between 39 and 29/DV5: 1.0\n",
      "Alignment score between 39 and 16: 0.0\n",
      "Alignment score between 39 and 30: 1.0\n",
      "Alignment score between 39 and 6: 0.0\n",
      "Alignment score between 39 and 23/DV6: 1.0\n",
      "Alignment score between 39 and 36/DV7: 1.0\n",
      "Alignment score between 39 and 8-2: 0.0\n",
      "Alignment score between 39 and 18: 0.0\n",
      "Alignment score between 39 and 34: 1.0\n",
      "Alignment score between 39 and 40: 0.0\n",
      "Alignment score between 39 and 7: 0.0\n",
      "Alignment score between 39 and 14/DV4*02: 0.0\n",
      "Alignment score between 39 and 8-6*02: 0.0\n",
      "Alignment score between 41 and 2: 0.0\n",
      "Alignment score between 41 and 9-1: 1.0\n",
      "Alignment score between 41 and 1-1: 1.0\n",
      "Alignment score between 41 and 8-6: 0.0\n",
      "Alignment score between 41 and 26-2: 0.0\n",
      "Alignment score between 41 and 24: 1.0\n",
      "Alignment score between 41 and 8-3: 0.0\n",
      "Alignment score between 41 and 17: 1.0\n",
      "Alignment score between 41 and 29/DV5: 0.0\n",
      "Alignment score between 41 and 16: 1.0\n",
      "Alignment score between 41 and 30: 0.0\n",
      "Alignment score between 41 and 6: 0.0\n",
      "Alignment score between 41 and 23/DV6: 0.0\n",
      "Alignment score between 41 and 36/DV7: 0.0\n",
      "Alignment score between 41 and 8-2: 0.0\n",
      "Alignment score between 41 and 18: 1.0\n",
      "Alignment score between 41 and 34: 1.0\n",
      "Alignment score between 41 and 40: 1.0\n",
      "Alignment score between 41 and 7: 0.0\n",
      "Alignment score between 41 and 14/DV4*02: 1.0\n",
      "Alignment score between 41 and 8-6*02: 0.0\n",
      "Alignment score between 2 and 9-1: 0.0\n",
      "Alignment score between 2 and 1-1: 0.0\n",
      "Alignment score between 2 and 8-6: 0.0\n",
      "Alignment score between 2 and 26-2: 1.0\n",
      "Alignment score between 2 and 24: 1.0\n",
      "Alignment score between 2 and 8-3: 0.0\n",
      "Alignment score between 2 and 17: 0.0\n",
      "Alignment score between 2 and 29/DV5: 1.0\n",
      "Alignment score between 2 and 16: 0.0\n",
      "Alignment score between 2 and 30: 0.0\n",
      "Alignment score between 2 and 6: 0.0\n",
      "Alignment score between 2 and 23/DV6: 1.0\n",
      "Alignment score between 2 and 36/DV7: 0.0\n",
      "Alignment score between 2 and 8-2: 1.0\n",
      "Alignment score between 2 and 18: 0.0\n",
      "Alignment score between 2 and 34: 0.0\n",
      "Alignment score between 2 and 40: 0.0\n",
      "Alignment score between 2 and 7: 0.0\n",
      "Alignment score between 2 and 14/DV4*02: 1.0\n",
      "Alignment score between 2 and 8-6*02: 1.0\n",
      "Alignment score between 9-1 and 1-1: 2.0\n",
      "Alignment score between 9-1 and 8-6: 1.0\n",
      "Alignment score between 9-1 and 26-2: 1.0\n",
      "Alignment score between 9-1 and 24: 0.0\n",
      "Alignment score between 9-1 and 8-3: 1.0\n",
      "Alignment score between 9-1 and 17: 1.0\n",
      "Alignment score between 9-1 and 29/DV5: 1.0\n",
      "Alignment score between 9-1 and 16: 1.0\n",
      "Alignment score between 9-1 and 30: 0.0\n",
      "Alignment score between 9-1 and 6: 0.0\n",
      "Alignment score between 9-1 and 23/DV6: 0.0\n",
      "Alignment score between 9-1 and 36/DV7: 0.0\n",
      "Alignment score between 9-1 and 8-2: 1.0\n",
      "Alignment score between 9-1 and 18: 1.0\n",
      "Alignment score between 9-1 and 34: 0.0\n",
      "Alignment score between 9-1 and 40: 0.0\n",
      "Alignment score between 9-1 and 7: 0.0\n",
      "Alignment score between 9-1 and 14/DV4*02: 1.0\n",
      "Alignment score between 9-1 and 8-6*02: 1.0\n",
      "Alignment score between 1-1 and 8-6: 1.0\n",
      "Alignment score between 1-1 and 26-2: 1.0\n",
      "Alignment score between 1-1 and 24: 0.0\n",
      "Alignment score between 1-1 and 8-3: 1.0\n",
      "Alignment score between 1-1 and 17: 1.0\n",
      "Alignment score between 1-1 and 29/DV5: 0.0\n",
      "Alignment score between 1-1 and 16: 1.0\n",
      "Alignment score between 1-1 and 30: 0.0\n",
      "Alignment score between 1-1 and 6: 0.0\n",
      "Alignment score between 1-1 and 23/DV6: 0.0\n",
      "Alignment score between 1-1 and 36/DV7: 0.0\n",
      "Alignment score between 1-1 and 8-2: 1.0\n",
      "Alignment score between 1-1 and 18: 1.0\n",
      "Alignment score between 1-1 and 34: 0.0\n",
      "Alignment score between 1-1 and 40: 0.0\n",
      "Alignment score between 1-1 and 7: 0.0\n",
      "Alignment score between 1-1 and 14/DV4*02: 1.0\n",
      "Alignment score between 1-1 and 8-6*02: 1.0\n",
      "Alignment score between 8-6 and 26-2: 1.0\n",
      "Alignment score between 8-6 and 24: 0.0\n",
      "Alignment score between 8-6 and 8-3: 2.0\n",
      "Alignment score between 8-6 and 17: 0.0\n",
      "Alignment score between 8-6 and 29/DV5: 0.0\n",
      "Alignment score between 8-6 and 16: 1.0\n",
      "Alignment score between 8-6 and 30: 0.0\n",
      "Alignment score between 8-6 and 6: 1.0\n",
      "Alignment score between 8-6 and 23/DV6: 1.0\n",
      "Alignment score between 8-6 and 36/DV7: 1.0\n",
      "Alignment score between 8-6 and 8-2: 2.0\n",
      "Alignment score between 8-6 and 18: 1.0\n",
      "Alignment score between 8-6 and 34: 0.0\n",
      "Alignment score between 8-6 and 40: 0.0\n",
      "Alignment score between 8-6 and 7: 0.0\n",
      "Alignment score between 8-6 and 14/DV4*02: 0.0\n",
      "Alignment score between 8-6 and 8-6*02: 3.0\n",
      "Alignment score between 26-2 and 24: 1.0\n",
      "Alignment score between 26-2 and 8-3: 1.0\n",
      "Alignment score between 26-2 and 17: 0.0\n",
      "Alignment score between 26-2 and 29/DV5: 1.0\n",
      "Alignment score between 26-2 and 16: 1.0\n",
      "Alignment score between 26-2 and 30: 0.0\n",
      "Alignment score between 26-2 and 6: 1.0\n",
      "Alignment score between 26-2 and 23/DV6: 2.0\n",
      "Alignment score between 26-2 and 36/DV7: 1.0\n",
      "Alignment score between 26-2 and 8-2: 2.0\n",
      "Alignment score between 26-2 and 18: 0.0\n",
      "Alignment score between 26-2 and 34: 0.0\n",
      "Alignment score between 26-2 and 40: 0.0\n",
      "Alignment score between 26-2 and 7: 0.0\n",
      "Alignment score between 26-2 and 14/DV4*02: 1.0\n",
      "Alignment score between 26-2 and 8-6*02: 2.0\n",
      "Alignment score between 24 and 8-3: 0.0\n",
      "Alignment score between 24 and 17: 0.0\n",
      "Alignment score between 24 and 29/DV5: 1.0\n",
      "Alignment score between 24 and 16: 0.0\n",
      "Alignment score between 24 and 30: 0.0\n",
      "Alignment score between 24 and 6: 0.0\n",
      "Alignment score between 24 and 23/DV6: 1.0\n",
      "Alignment score between 24 and 36/DV7: 0.0\n",
      "Alignment score between 24 and 8-2: 1.0\n",
      "Alignment score between 24 and 18: 0.0\n",
      "Alignment score between 24 and 34: 1.0\n",
      "Alignment score between 24 and 40: 1.0\n",
      "Alignment score between 24 and 7: 0.0\n",
      "Alignment score between 24 and 14/DV4*02: 1.0\n",
      "Alignment score between 24 and 8-6*02: 1.0\n",
      "Alignment score between 8-3 and 17: 0.0\n",
      "Alignment score between 8-3 and 29/DV5: 0.0\n",
      "Alignment score between 8-3 and 16: 0.0\n",
      "Alignment score between 8-3 and 30: 1.0\n",
      "Alignment score between 8-3 and 6: 0.0\n",
      "Alignment score between 8-3 and 23/DV6: 1.0\n",
      "Alignment score between 8-3 and 36/DV7: 1.0\n",
      "Alignment score between 8-3 and 8-2: 2.0\n",
      "Alignment score between 8-3 and 18: 1.0\n",
      "Alignment score between 8-3 and 34: 1.0\n",
      "Alignment score between 8-3 and 40: 0.0\n",
      "Alignment score between 8-3 and 7: 0.0\n",
      "Alignment score between 8-3 and 14/DV4*02: 0.0\n",
      "Alignment score between 8-3 and 8-6*02: 2.0\n",
      "Alignment score between 17 and 29/DV5: 0.0\n",
      "Alignment score between 17 and 16: 1.0\n",
      "Alignment score between 17 and 30: 0.0\n",
      "Alignment score between 17 and 6: 0.0\n",
      "Alignment score between 17 and 23/DV6: 0.0\n",
      "Alignment score between 17 and 36/DV7: 1.0\n",
      "Alignment score between 17 and 8-2: 0.0\n",
      "Alignment score between 17 and 18: 1.0\n",
      "Alignment score between 17 and 34: 0.0\n",
      "Alignment score between 17 and 40: 0.0\n",
      "Alignment score between 17 and 7: 1.0\n",
      "Alignment score between 17 and 14/DV4*02: 1.0\n",
      "Alignment score between 17 and 8-6*02: 0.0\n",
      "Alignment score between 29/DV5 and 16: 0.0\n",
      "Alignment score between 29/DV5 and 30: 0.0\n",
      "Alignment score between 29/DV5 and 6: 0.0\n",
      "Alignment score between 29/DV5 and 23/DV6: 4.0\n",
      "Alignment score between 29/DV5 and 36/DV7: 3.0\n",
      "Alignment score between 29/DV5 and 8-2: 1.0\n",
      "Alignment score between 29/DV5 and 18: 0.0\n",
      "Alignment score between 29/DV5 and 34: 0.0\n",
      "Alignment score between 29/DV5 and 40: 0.0\n",
      "Alignment score between 29/DV5 and 7: 0.0\n",
      "Alignment score between 29/DV5 and 14/DV4*02: 3.0\n",
      "Alignment score between 29/DV5 and 8-6*02: 1.0\n",
      "Alignment score between 16 and 30: 0.0\n",
      "Alignment score between 16 and 6: 1.0\n",
      "Alignment score between 16 and 23/DV6: 1.0\n",
      "Alignment score between 16 and 36/DV7: 1.0\n",
      "Alignment score between 16 and 8-2: 0.0\n",
      "Alignment score between 16 and 18: 1.0\n",
      "Alignment score between 16 and 34: 0.0\n",
      "Alignment score between 16 and 40: 0.0\n",
      "Alignment score between 16 and 7: 0.0\n",
      "Alignment score between 16 and 14/DV4*02: 1.0\n",
      "Alignment score between 16 and 8-6*02: 1.0\n",
      "Alignment score between 30 and 6: 0.0\n",
      "Alignment score between 30 and 23/DV6: 1.0\n",
      "Alignment score between 30 and 36/DV7: 1.0\n",
      "Alignment score between 30 and 8-2: 0.0\n",
      "Alignment score between 30 and 18: 0.0\n",
      "Alignment score between 30 and 34: 1.0\n",
      "Alignment score between 30 and 40: 1.0\n",
      "Alignment score between 30 and 7: 0.0\n",
      "Alignment score between 30 and 14/DV4*02: 1.0\n",
      "Alignment score between 30 and 8-6*02: 1.0\n",
      "Alignment score between 6 and 23/DV6: 1.0\n",
      "Alignment score between 6 and 36/DV7: 1.0\n",
      "Alignment score between 6 and 8-2: 0.0\n",
      "Alignment score between 6 and 18: 0.0\n",
      "Alignment score between 6 and 34: 0.0\n",
      "Alignment score between 6 and 40: 0.0\n",
      "Alignment score between 6 and 7: 0.0\n",
      "Alignment score between 6 and 14/DV4*02: 0.0\n",
      "Alignment score between 6 and 8-6*02: 1.0\n",
      "Alignment score between 23/DV6 and 36/DV7: 4.0\n",
      "Alignment score between 23/DV6 and 8-2: 1.0\n",
      "Alignment score between 23/DV6 and 18: 0.0\n",
      "Alignment score between 23/DV6 and 34: 1.0\n",
      "Alignment score between 23/DV6 and 40: 0.0\n",
      "Alignment score between 23/DV6 and 7: 0.0\n",
      "Alignment score between 23/DV6 and 14/DV4*02: 3.0\n",
      "Alignment score between 23/DV6 and 8-6*02: 1.0\n",
      "Alignment score between 36/DV7 and 8-2: 0.0\n",
      "Alignment score between 36/DV7 and 18: 0.0\n",
      "Alignment score between 36/DV7 and 34: 1.0\n",
      "Alignment score between 36/DV7 and 40: 0.0\n",
      "Alignment score between 36/DV7 and 7: 1.0\n",
      "Alignment score between 36/DV7 and 14/DV4*02: 3.0\n",
      "Alignment score between 36/DV7 and 8-6*02: 1.0\n",
      "Alignment score between 8-2 and 18: 1.0\n",
      "Alignment score between 8-2 and 34: 0.0\n",
      "Alignment score between 8-2 and 40: 0.0\n",
      "Alignment score between 8-2 and 7: 0.0\n",
      "Alignment score between 8-2 and 14/DV4*02: 1.0\n",
      "Alignment score between 8-2 and 8-6*02: 3.0\n",
      "Alignment score between 18 and 34: 0.0\n",
      "Alignment score between 18 and 40: 0.0\n",
      "Alignment score between 18 and 7: 0.0\n",
      "Alignment score between 18 and 14/DV4*02: 1.0\n",
      "Alignment score between 18 and 8-6*02: 1.0\n",
      "Alignment score between 34 and 40: 1.0\n",
      "Alignment score between 34 and 7: 0.0\n",
      "Alignment score between 34 and 14/DV4*02: 1.0\n",
      "Alignment score between 34 and 8-6*02: 0.0\n",
      "Alignment score between 40 and 7: 0.0\n",
      "Alignment score between 40 and 14/DV4*02: 2.0\n",
      "Alignment score between 40 and 8-6*02: 1.0\n",
      "Alignment score between 7 and 14/DV4*02: 0.0\n",
      "Alignment score between 7 and 8-6*02: 0.0\n",
      "Alignment score between 14/DV4*02 and 8-6*02: 3.0\n"
     ]
    }
   ],
   "source": [
    "from Bio import pairwise2\n",
    "\n",
    "# Example TRAV gene segments\n",
    "sequences = ['12-2', '12-1', '12-3', '13-1', '8-4', '8-1', '14/DV4', '4', '5', '19', '9-2', '22', '25', '10', '39', '41', '2', '9-1', '1-1', '8-6', '26-2', '24', '8-3', '17', '29/DV5', '16', '30', '6', '23/DV6', '36/DV7', '8-2', '18', '34', '40', '7', '14/DV4*02', '8-6*02']\n",
    "\n",
    "\n",
    "# Compute pairwise alignment scores using Needleman-Wunsch\n",
    "align_scores = {}\n",
    "for i, seq1 in enumerate(sequences):\n",
    "    for j, seq2 in enumerate(sequences):\n",
    "        if j > i:\n",
    "            align_score = pairwise2.align.globalxx(seq1, seq2, score_only=True)\n",
    "            align_scores[(i, j)] = align_score\n",
    "            align_scores[(j, i)] = align_score\n",
    "\n",
    "# Print the alignment scores\n",
    "for i, seq1 in enumerate(sequences):\n",
    "    for j, seq2 in enumerate(sequences):\n",
    "        if j > i:\n",
    "            print(f\"Alignment score between {seq1} and {seq2}: {align_scores[(i, j)]}\")"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'12-2': 44.0, '12-1': 43.0, '12-3': 45.0, '13-1': 38.0, '8-4': 29.0, '8-1': 33.0, '14/DV4': 39.0, '4': 8.0, '5': 3.0, '19': 19.0, '9-2': 30.0, '22': 17.0, '25': 17.0, '10': 20.0, '39': 13.0, '41': 21.0, '2': 14.0, '9-1': 31.0, '1-1': 32.0, '8-6': 27.0, '26-2': 36.0, '24': 21.0, '8-3': 28.0, '17': 18.0, '29/DV5': 35.0, '16': 22.0, '30': 13.0, '6': 7.0, '23/DV6': 41.0, '36/DV7': 32.0, '8-2': 33.0, '18': 21.0, '34': 16.0, '40': 13.0, '7': 3.0, '14/DV4*02': 57.0, '8-6*02': 45.0}\n"
     ]
    }
   ],
   "source": [
    "from Bio import pairwise2\n",
    "\n",
    "def get_similarity_score(seq1, seq2):\n",
    "    alignments = pairwise2.align.globalxx(seq1, seq2)\n",
    "    alignment_scores = [a[2] for a in alignments]\n",
    "    return max(alignment_scores, default=0)\n",
    "\n",
    "sequences = ['12-2', '12-1', '12-3', '13-1', '8-4', '8-1', '14/DV4', '4', '5', '19', '9-2', '22', '25', '10', '39', '41', '2', '9-1', '1-1', '8-6', '26-2', '24', '8-3', '17', '29/DV5', '16', '30', '6', '23/DV6', '36/DV7', '8-2', '18', '34', '40', '7', '14/DV4*02', '8-6*02']\n",
    "\n",
    "scores = {}\n",
    "for seq in sequences:\n",
    "    score = sum(get_similarity_score(seq, other_seq) for other_seq in sequences)\n",
    "    scores[seq] = score\n",
    "\n",
    "print(scores)\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "TRAV26-1*01: 0.47022136752577504\n",
      "TRAV20*01: 0.47841808883725045\n",
      "TRAV38-2/DV8*01: 0.5862872286618203\n",
      "TRAV26-1*01: 0.47022136752577504\n",
      "TRAV20*01: 0.47841808883725045\n",
      "TRAV26-1*01: 0.47022136752577504\n",
      "TRAV20*01: 0.47841808883725045\n",
      "TRAV26-1*01: 0.47022136752577504\n",
      "TRAV20*01: 0.47841808883725045\n",
      "TRAV38-1*01: 0.4963156578460396\n",
      "TRAV12-2*01: 0.23251644949298814\n",
      "TRAV12-2*01: 0.23251644949298814\n",
      "TRAV12-2*01: 0.23251644949298814\n",
      "TRAV12-2*01: 0.23251644949298814\n",
      "TRAV12-2*01: 0.23251644949298814\n",
      "TRAV12-2*01: 0.23251644949298814\n",
      "TRAV12-2*01: 0.23251644949298814\n",
      "TRAV12-2*01: 0.23251644949298814\n",
      "TRAV12-2*01: 0.23251644949298814\n",
      "TRAV12-2*01: 0.23251644949298814\n",
      "TRAV12-2*01: 0.23251644949298814\n",
      "TRAV12-2*01: 0.23251644949298814\n",
      "TRAV12-2*01: 0.23251644949298814\n",
      "TRAV12-2*01: 0.23251644949298814\n",
      "TRAV12-2*01: 0.23251644949298814\n",
      "TRAV3*01: 0.47841808883725045\n",
      "TRAV12-1*01: 0.23251644949298814\n",
      "TRAV27*01: 0.5030082527716767\n",
      "TRAV1-2*01: 0.5030082527716767\n",
      "TRAV35*01: 0.5030082527716767\n",
      "TRAV13-2*01: 0.4538279249028242\n",
      "TRAV12-2*01: 0.23251644949298814\n",
      "TRAV12-2*01: 0.23251644949298814\n",
      "TRAV12-2*01: 0.23251644949298814\n",
      "TRAV12-2*01: 0.23251644949298814\n",
      "TRAV12-2*01: 0.23251644949298814\n",
      "TRAV12-2*01: 0.23251644949298814\n",
      "TRAV12-2*01: 0.23251644949298814\n",
      "TRAV12-2*01: 0.23251644949298814\n",
      "TRAV3*01: 0.47841808883725045\n",
      "TRAV12-2*01: 0.23251644949298814\n",
      "TRAV12-2*01: 0.23251644949298814\n",
      "TRAV12-2*01: 0.23251644949298814\n",
      "TRAV12-2*01: 0.23251644949298814\n",
      "TRAV12-2*01: 0.23251644949298814\n",
      "TRAV12-2*01: 0.23251644949298814\n",
      "TRAV12-2*01: 0.23251644949298814\n",
      "TRAV12-3*01: 0.23251644949298814\n",
      "TRAV12-3*01: 0.23251644949298814\n",
      "TRAV12-3*01: 0.23251644949298814\n",
      "TRAV13-1*01: 0.4538279249028242\n",
      "TRAV13-1*01: 0.4538279249028242\n",
      "TRAV13-1*01: 0.4538279249028242\n",
      "TRAV13-1*01: 0.4538279249028242\n",
      "TRAV13-1*01: 0.4538279249028242\n",
      "TRAV13-1*01: 0.4538279249028242\n",
      "TRAV8-4*01: 0.4948115314602013\n",
      "TRAV8-1*01: 0.4948115314602013\n",
      "TRAV3*01: 0.47841808883725045\n",
      "TRAV14/DV4*01: 0.5929798235874575\n",
      "TRAV3*01: 0.47841808883725045\n",
      "TRAV26-1*01: 0.47022136752577504\n"
     ]
    }
   ],
   "source": [
    "'''\n",
    "from sklearn.feature_extraction.text import CountVectorizer\n",
    "from sklearn.metrics.pairwise import pairwise_distances\n",
    "import scipy.spatial.distance as dist\n",
    "\n",
    "# Create a count vectorizer to convert the strings into binary term frequencies\n",
    "vectorizer = CountVectorizer(binary=True)\n",
    "X = vectorizer.fit_transform(sequences)\n",
    "\n",
    "dense_matrix = X.toarray()\n",
    "\n",
    "# Compute pairwise distances between rows\n",
    "distances = dist.pdist(dense_matrix, metric='cosine')\n",
    "square_distances = dist.squareform(distances)\n",
    "\n",
    "# Calculate the score for each string as the average similarity with the other strings\n",
    "scores = []\n",
    "for i in range(len(sequences)):\n",
    "    score = sum(square_distances[i]) / (len(sequences) - 1)\n",
    "    scores.append(score)\n",
    "\n",
    "# Print the scores for each string\n",
    "for i in range(len(sequences)):\n",
    "    print(f\"{sequences[i]}: {scores[i]}\")\n",
    "'''"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "fetching data stage\n",
      "              cdr3_a_aa         v_a_gene   j_a_gene             cdr3_b_aa  \\\n",
      "0         CIVRAPGRADMRF      TRAV26-1*01  TRAJ43*01  CASSYLPGQGDHYSNQPQHF   \n",
      "2        CAVPSGAGSYQLTF        TRAV20*01  TRAJ28*01   CASSFEPGQGFYSNQPQHF   \n",
      "4         CAYRPPGTYKYIF  TRAV38-2/DV8*01  TRAJ40*01        CASSALASLNEQFF   \n",
      "5         CIVRAPGRADMRF      TRAV26-1*01  TRAJ43*01  CASSYLPGQGDHYSNQPQHF   \n",
      "7        CAVPSGAGSYQLTF        TRAV20*01  TRAJ28*01   CASSFEPGQGFYSNQPQHF   \n",
      "...                 ...              ...        ...                   ...   \n",
      "59806      CAVYNQGGKLIF       TRAV8-1*01  TRAJ23*01      CASSALVAFSTDTQYF   \n",
      "59807    CAVRALYNFNKFYF         TRAV3*01  TRAJ21*01       CASSSLAGEVNEQFF   \n",
      "59808     CAVESYNTDKLIF        TRAV41*01  TRAJ34*01   CASSSYTQRGLDYTGELFF   \n",
      "59810    CLVGENSGGFKTIF         TRAV4*01   TRAJ9*01                   NaN   \n",
      "59811  CAMREGQGNAGNMLTF    TRAV14/DV4*01  TRAJ39*01     CASSQDYWGGGADTQYF   \n",
      "\n",
      "         v_b_gene d.beta    j_b_gene      species        mhc.a mhc.b  ...  \\\n",
      "0       TRBV13*01    NaN  TRBJ1-5*01  HomoSapiens     HLA-B*08   B2M  ...   \n",
      "2       TRBV13*01    NaN  TRBJ1-5*01  HomoSapiens     HLA-B*08   B2M  ...   \n",
      "4       TRBV14*01    NaN  TRBJ2-1*01  HomoSapiens     HLA-B*08   B2M  ...   \n",
      "5       TRBV13*01    NaN  TRBJ1-5*01  HomoSapiens     HLA-B*08   B2M  ...   \n",
      "7       TRBV13*01    NaN  TRBJ1-5*01  HomoSapiens     HLA-B*08   B2M  ...   \n",
      "...           ...    ...         ...          ...          ...   ...  ...   \n",
      "59806  TRBV6-4*01    NaN  TRBJ2-3*01  HomoSapiens  HLA-A*24:01   B2M  ...   \n",
      "59807  TRBV7-2*01    NaN  TRBJ2-1*01  HomoSapiens  HLA-A*24:01   B2M  ...   \n",
      "59808   TRBV27*01    NaN  TRBJ2-2*01  HomoSapiens  HLA-A*24:01   B2M  ...   \n",
      "59810         NaN    NaN         NaN  HomoSapiens  HLA-A*24:01   B2M  ...   \n",
      "59811  TRBV4-1*01    NaN  TRBJ2-3*01  HomoSapiens  HLA-A*24:01   B2M  ...   \n",
      "\n",
      "      meta.replica.id meta.clone.id meta.epitope.id  meta.tissue  \\\n",
      "0                 NaN           NaN             NaN         PBMC   \n",
      "2                 NaN           NaN             NaN         PBMC   \n",
      "4                 NaN           NaN             NaN         PBMC   \n",
      "5                 NaN           NaN             NaN  TCL culture   \n",
      "7                 NaN           NaN             NaN  TCL culture   \n",
      "...               ...           ...             ...          ...   \n",
      "59806             NaN           NaN           S1208         PBMC   \n",
      "59807             NaN           NaN           S1208         PBMC   \n",
      "59808             NaN           NaN           S1208         PBMC   \n",
      "59810             NaN           NaN           S1208         PBMC   \n",
      "59811             NaN           NaN           S1208         PBMC   \n",
      "\n",
      "                                          meta.donor.MHC  \\\n",
      "0      HLA-A*02:01,HLA-A*24:02;HLA-B*08:01,HLA-B*5701...   \n",
      "2      HLA-A*02,HLA-A*24:02;HLA-B*08:01,HLA-B*15;HLA-...   \n",
      "4      HLA-A*01:01,HLA-A*02:01;HLA-B*08:01,HLA-B*57:0...   \n",
      "5      HLA-A*02:01,HLA-A*24:02;HLA-B*08:01,HLA-B*5701...   \n",
      "7      HLA-A*02,HLA-A*24:02;HLA-B*08:01,HLA-B*15;HLA-...   \n",
      "...                                                  ...   \n",
      "59806                                                NaN   \n",
      "59807                                                NaN   \n",
      "59808                                                NaN   \n",
      "59810                                                NaN   \n",
      "59811                                                NaN   \n",
      "\n",
      "      meta.donor.MHC.method meta.structure.id  \\\n",
      "0                       NaN               NaN   \n",
      "2                       NaN               NaN   \n",
      "4                       NaN               NaN   \n",
      "5                       NaN               NaN   \n",
      "7                       NaN               NaN   \n",
      "...                     ...               ...   \n",
      "59806                   NaN               NaN   \n",
      "59807                   NaN               NaN   \n",
      "59808                   NaN               NaN   \n",
      "59810                   NaN               NaN   \n",
      "59811                   NaN               NaN   \n",
      "\n",
      "                                           cdr3fix.alpha  \\\n",
      "0      {\"cdr3\": \"CIVRAPGRADMRF\", \"cdr3_old\": \"CIVRAPG...   \n",
      "2      {\"cdr3\": \"CAVPSGAGSYQLTF\", \"cdr3_old\": \"CAVPSG...   \n",
      "4      {\"cdr3\": \"CAYRPPGTYKYIF\", \"cdr3_old\": \"CAYRPPG...   \n",
      "5      {\"cdr3\": \"CIVRAPGRADMRF\", \"cdr3_old\": \"CIVRAPG...   \n",
      "7      {\"cdr3\": \"CAVPSGAGSYQLTF\", \"cdr3_old\": \"CAVPSG...   \n",
      "...                                                  ...   \n",
      "59806  {\"cdr3\": \"CAVYNQGGKLIF\", \"cdr3_old\": \"CAVYNQGG...   \n",
      "59807  {\"cdr3\": \"CAVRALYNFNKFYF\", \"cdr3_old\": \"CAVRAL...   \n",
      "59808  {\"cdr3\": \"CAVESYNTDKLIF\", \"cdr3_old\": \"CAVESYN...   \n",
      "59810  {\"cdr3\": \"CLVGENSGGFKTIF\", \"cdr3_old\": \"CLVGEN...   \n",
      "59811  {\"cdr3\": \"CAMREGQGNAGNMLTF\", \"cdr3_old\": \"CAMR...   \n",
      "\n",
      "                                            cdr3fix.beta vdjdb.score  \n",
      "0      {\"cdr3\": \"CASSYLPGQGDHYSNQPQHF\", \"cdr3_old\": \"...           2  \n",
      "2      {\"cdr3\": \"CASSFEPGQGFYSNQPQHF\", \"cdr3_old\": \"C...           2  \n",
      "4      {\"cdr3\": \"CASSALASLNEQFF\", \"cdr3_old\": \"CASSAL...           2  \n",
      "5      {\"cdr3\": \"CASSYLPGQGDHYSNQPQHF\", \"cdr3_old\": \"...           2  \n",
      "7      {\"cdr3\": \"CASSFEPGQGFYSNQPQHF\", \"cdr3_old\": \"C...           2  \n",
      "...                                                  ...         ...  \n",
      "59806  {\"cdr3\": \"CASSALVAFSTDTQYF\", \"cdr3_old\": \"CASS...           1  \n",
      "59807  {\"cdr3\": \"CASSSLAGEVNEQFF\", \"cdr3_old\": \"CASSS...           1  \n",
      "59808  {\"cdr3\": \"CASSSYTQRGLDYTGELFF\", \"cdr3_old\": \"C...           1  \n",
      "59810                                                NaN           1  \n",
      "59811  {\"cdr3\": \"CASSQDYWGGGADTQYF\", \"cdr3_old\": \"CAS...           1  \n",
      "\n",
      "[33329 rows x 34 columns]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/liuqianyi/Mini-Project-TB2-PA/src/preprocess/feature_filter.py:4: DtypeWarning: Columns (20,29,30) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  df = pd.read_table('../../data/vdjdb_full.tsv').rename(\n"
     ]
    }
   ],
   "source": [
    "from src.preprocess.setup import config\n",
    "from src.preprocess.feature_filter import load_data\n",
    "from src.preprocess.label import generate_label\n",
    "import src.preprocess.feature_filter\n",
    "\n",
    "config.setConfig(config.speciesType.human, config.chainType.alpha)\n",
    "data = load_data()\n",
    "print(data)"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['TRAV26-1*01' 'TRAV20*01' 'TRAV38-2/DV8*01' 'TRAV38-1*01' 'TRAV12-2*01'\n",
      " 'TRAV3*01' 'TRAV12-1*01' 'TRAV27*01' 'TRAV1-2*01' 'TRAV35*01'\n",
      " 'TRAV13-2*01' 'TRAV21*01' 'TRAV12-3*01' 'TRAV13-1*01' 'TRAV8-4*01'\n",
      " 'TRAV8-1*01' 'TRAV14/DV4*01' 'TRAV4*01' 'TRAV5*01' 'TRAV19*01'\n",
      " 'TRAV9-2*01' 'TRAV22*01' 'TRAV25*01' 'TRAV10*01' 'TRAV39*01' 'TRAV41*01'\n",
      " 'TRAV2*01' 'TRAV9-1*01' 'TRAV1-1*01' 'TRAV8-6*01' 'TRAV26-2*01'\n",
      " 'TRAV24*01' 'TRAV8-3*01' 'TRAV17*01' 'TRAV29/DV5*01' 'TRAV16*01'\n",
      " 'TRAV30*01' 'TRAV6*01' 'TRAV23/DV6*01' 'TRAV36/DV7*01' 'TRAV8-2*01'\n",
      " 'TRAV18*01' 'TRAV34*01' 'TRAV40*01' 'TRAV7*01' 'TRAV14/DV4*02'\n",
      " 'TRAV8-6*02']\n",
      "['TRAJ43*01' 'TRAJ28*01' 'TRAJ40*01' 'TRAJ48*01' 'TRAJ42*01' 'TRAJ45*01'\n",
      " 'TRAJ23*01' 'TRAJ31*01' 'TRAJ53*01' 'TRAJ39*01' 'TRAJ7*01' 'TRAJ3*01'\n",
      " 'TRAJ20*01' 'TRAJ9*01' 'TRAJ33*01' 'TRAJ22*01' 'TRAJ37*01' 'TRAJ47*01'\n",
      " 'TRAJ34*01' 'TRAJ8*01' 'TRAJ57*01' 'TRAJ21*01' 'TRAJ54*01' 'TRAJ49*01'\n",
      " 'TRAJ44*01' 'TRAJ30*01' 'TRAJ10*01' 'TRAJ52*01' 'TRAJ15*01' 'TRAJ17*01'\n",
      " 'TRAJ6*01' 'TRAJ11*01' 'TRAJ56*01' 'TRAJ29*01' 'TRAJ13*01' 'TRAJ5*01'\n",
      " 'TRAJ24*01' 'TRAJ12*01' 'TRAJ27*01' 'TRAJ18*01' 'TRAJ32*01' 'TRAJ4*01'\n",
      " 'TRAJ36*01' 'TRAJ50*01' 'TRAJ41*01' 'TRAJ26*01' 'TRAJ38*01' 'TRAJ46*01'\n",
      " 'TRAJ13*02' 'TRAJ24*02' 'TRAJ14*01']\n",
      "['26-1', '20', '38-2/DV8', '38-1', '12-2', '3', '12-1', '27', '1-2', '35', '13-2', '21', '12-3', '13-1', '8-4', '8-1', '14/DV4', '4', '5', '19', '9-2', '22', '25', '10', '39', '41', '2', '9-1', '1-1', '8-6', '26-2', '24', '8-3', '17', '29/DV5', '16', '30', '6', '23/DV6', '36/DV7', '8-2', '18', '34', '40', '7', '14/DV4*02', '8-6*02']\n"
     ]
    },
    {
     "data": {
      "text/plain": "'\\nscores = {}\\nfor seq in sequences:\\n    score = sum(get_similarity_score(seq, other_seq) for other_seq in sequences)\\n    scores[seq] = score\\n\\nprint(scores)\\n\\n'"
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import numpy as np\n",
    "x = data['v_a_gene'].unique()\n",
    "print(x)\n",
    "\n",
    "j_a_gene = data['j_a_gene'].unique()\n",
    "print(j_a_gene)\n",
    "\n",
    "\n",
    "\n",
    "output_list = [s.replace('TRAV', '').replace('*01', '') for s in x]\n",
    "\n",
    "print(output_list)\n",
    "\n",
    "'''\n",
    "scores = {}\n",
    "for seq in sequences:\n",
    "    score = sum(get_similarity_score(seq, other_seq) for other_seq in sequences)\n",
    "    scores[seq] = score\n",
    "\n",
    "print(scores)\n",
    "\n",
    "'''"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "outputs": [],
   "source": [],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "outputs": [],
   "source": [
    "'''\n",
    "from Bio.Align.Applications import MuscleCommandline\n",
    "from io import StringIO\n",
    "from Bio import AlignIO\n",
    "\n",
    "# Extract V column from sequences list\n",
    "sequences_V = [s.split('\\t')[0] for s in sequences]\n",
    "\n",
    "# Write sequences to a temporary file\n",
    "temp_seq_file = 'sequences.fasta'\n",
    "with open(temp_seq_file, 'w') as f:\n",
    "    for i, seq in enumerate(sequences_V):\n",
    "        f.write('>seq{}\\n{}\\n'.format(i, seq))\n",
    "\n",
    "# Run MUSCLE alignment\n",
    "muscle_exe = '/Users/liuqianyi/Documents/muscle/muscle.macos_intel64'\n",
    "muscle_cline = MuscleCommandline(muscle_exe, input=temp_seq_file)\n",
    "stdout, stderr = muscle_cline()\n",
    "align = AlignIO.read(StringIO(stdout), 'fasta')\n",
    "\n",
    "# Print similarity scores\n",
    "for i, record1 in enumerate(align):\n",
    "    for j, record2 in enumerate(align):\n",
    "        score = align[:, i].similarity(align[:, j])\n",
    "        print('{} vs {}: {}'.format(record1.id, record2.id, score))\n",
    "'''\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "('12-1', '1-1', 86)\n",
      "('13-1', '1-1', 86)\n",
      "('8-1', '18', 80)\n",
      "('14/DV4', '14/DV4*02', 80)\n",
      "('19', '9-1', 80)\n",
      "('12-2', '12-3', 75)\n",
      "('12-2', '26-2', 75)\n",
      "('12-1', '13-1', 75)\n",
      "('12-3', '13-1', 75)\n",
      "('12-2', '22', 67)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/liuqianyi/Mini-Project-TB2-PA/venv/lib/python3.11/site-packages/fuzzywuzzy/fuzz.py:11: UserWarning: Using slow pure-python SequenceMatcher. Install python-Levenshtein to remove this warning\n",
      "  warnings.warn('Using slow pure-python SequenceMatcher. Install python-Levenshtein to remove this warning')\n"
     ]
    }
   ],
   "source": [
    "# Select the columns of interest\n",
    "columns = ['26-1', '20', '38-2/DV8', '38-1', '12-2', '3', '12-1', '27', '1-2', '35', '13-2', '21', '12-3', '13-1', '8-4', '8-1', '14/DV4', '4', '5', '19', '9-2', '22', '25', '10', '39', '41', '2', '9-1', '1-1', '8-6', '26-2', '24', '8-3', '17', '29/DV5', '16', '30', '6', '23/DV6', '36/DV7', '8-2', '18', '34', '40', '7', '14/DV4*02', '8-6*02']\n",
    "from fuzzywuzzy import fuzz\n",
    "\n",
    "columns = ['12-2', '12-1', '12-3', '13-1', '8-4', '8-1', '14/DV4', '4', '5', '19', '9-2', '22', '25', '10', '39', '41', '2', '9-1', '1-1', '8-6', '26-2', '24', '8-3', '17', '29/DV5', '16', '30', '6', '23/DV6', '36/DV7', '8-2', '18', '34', '40', '7', '14/DV4*02', '8-6*02']\n",
    "\n",
    "# Compute the similarity score for each pair of strings\n",
    "similarity_scores = []\n",
    "for i in range(len(columns)):\n",
    "    for j in range(i+1, len(columns)):\n",
    "        score = fuzz.token_sort_ratio(columns[i], columns[j])\n",
    "        similarity_scores.append((columns[i], columns[j], score))\n",
    "\n",
    "# Sort the scores in descending order\n",
    "similarity_scores.sort(key=lambda x: x[2], reverse=True)\n",
    "\n",
    "# Print the top 10 pairs of strings with the highest similarity scores\n",
    "for i in range(10):\n",
    "    print(similarity_scores[i])"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "outputs": [],
   "source": [],
   "metadata": {
    "collapsed": false
   }
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
